redis_session.c

```diff
@@ -183,6 +186,130 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+                lock_status->is_locked = 1;
+          } else if (max_lock_retries == -1 || i_lock_retry < max_lock_retries) {
+              usleep(lock_wait_time);
+          }
+        }
+        lock_status->is_locked = 1;
+
+        efree(cmd);
+        efree(response);
+    }
+}
+
+void lock_release(RedisSock *redis_sock, redis_session_lock_status *lock_status)
+{
+    if (lock_status->is_locked) {
+        char *cmd, *response;
+        int response_len, cmd_len;
+
+        upload_lock_release_script(redis_sock);
+        cmd_len = REDIS_SPPRINTF(&cmd, "EVALSHA", "sdss", REDIS_G(lock_release_lua_script_hash), strlen(REDIS_G(lock_release_lua_script_hash)), 1, lock_status->lock_key, strlen(lock_status->lock_key), lock_status->lock_secret_hash, strlen(lock_status->lock_secret_hash));
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+        response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+
+        // in case of redis script cache has been flushed
+        if (response == NULL) {
+            REDIS_G(lock_release_lua_script_uploaded) = 0;
+            upload_lock_release_script(redis_sock);
+            redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+            response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+        }
+
+        efree(cmd);
+        efree(response);
+    }
+}
+
+void upload_lock_release_script(RedisSock *redis_sock)
+{
+    if (!REDIS_G(lock_release_lua_script_uploaded)) {
+        char *cmd, *response, *release_script;
+        int response_len, cmd_len;
+        release_script = "if redis.call(\"get\",KEYS[1]) == ARGV[1] then return redis.call(\"del\",KEYS[1]) else return 0 end";
+
+        cmd_len = REDIS_SPPRINTF(&cmd, "SCRIPT", "ss", "LOAD", strlen("LOAD"), release_script, strlen(release_script));
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+        response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+        memset(REDIS_G(lock_release_lua_script_hash), 0, 41);
+        strncpy(REDIS_G(lock_release_lua_script_hash), response, strlen(response));
+
+        REDIS_G(lock_release_lua_script_uploaded) = 1;
+
+        efree(cmd);
+        efree(response);
+    }
+}
+
+void calculate_secret_hash(redis_session_lock_status *lock_status)
+{
+    int random_number, pid;
+    char hostname[64] = {0}, lock_secret_hash[41] = {0}, sha_digest[20];
+    PHP_SHA1_CTX sha_context;
+    smart_string lock_secret = {0};
+
+    gethostname(hostname, 64);
+
+    // Concatenating the redis lock secret
+    smart_string_appendl(&lock_secret, hostname, strlen(hostname));
+    smart_string_appendc(&lock_secret, '|');
+    smart_string_append_long(&lock_secret, getpid());
+
+    // Hashing the redis lock secret (value)
+    PHP_SHA1Init(&sha_context);
+    PHP_SHA1Update(&sha_context, lock_secret.c, lock_secret.len);
+    PHP_SHA1Final(sha_digest, &sha_context);
+    make_sha1_digest(lock_secret_hash, sha_digest);
```

Comment:
You are right, its not necessary anymore. In the beginning also a random number was part of a secret, which was the reason for hashing.
Will be removed.